csrc/preseg_passes/move_gather.cpp

```diff
@@ -0,0 +1,250 @@
+ */
+// clang-format on
+#include <preseg_passes/move_gather.h>
+
+#include <expr_simplifier.h>
+#include <fusion.h>
+#include <ir/builder.h>
+#include <ir/interface_nodes.h>
+#include <ir/internal_base_nodes.h>
+#include <ir/utils.h>
+#include <ops/alias.h>
+#include <ops/arith.h>
+#include <ops/indexing.h>
+#include <ops/utils.h>
+#include <scheduler/utils.h>
+#include <transform_replay.h>
+
+namespace nvfuser::preseg_passes {
+
+namespace {
+
+bool isUnaryPointwiseOp(const Expr* expr) {
+  // Check if the expression is a UnaryOp and a pointwise operation
+  return expr->isA<UnaryOp>() && ir_utils::isPointwiseTvOp(expr);
+}
+
+// For now we allow unary pointwise ops like cast/neg or a squeeze.
+// We only consider squeeze which has a single dim squeezed.
+std::optional<Expr*> getAllowedLookupTvDef(const GatherOp* gather_op) {
+  if (gather_op->lookupTv()->isFusionInput()) {
+    return std::nullopt;
+  }
+
+  auto producing_expr = gather_op->lookupTv()->definition();
+  NVF_ERROR(
+      producing_expr != nullptr, "Def for Lookup tensor view is nullptr.");
+
+  if (isUnaryPointwiseOp(producing_expr)) {
+    return producing_expr;
+  }
+
+  // We handle squeeze op where only one dim is squeezed.
+  if (producing_expr->isA<SqueezeOp>()) {
+    auto squeeze_op = producing_expr->as<SqueezeOp>();
+    if (std::count(
+            squeeze_op->getSqueezeDimFlags().begin(),
+            squeeze_op->getSqueezeDimFlags().end(),
+            true) == 1) {
+      return producing_expr;
+    }
+  }
+
+  return std::nullopt;
+}
+
+int64_t positionOfSqueezedDim(const std::vector<bool>& squeeze_dim_flags) {
+  auto it = std::find(squeeze_dim_flags.begin(), squeeze_dim_flags.end(), true);
+  return it != squeeze_dim_flags.end()
+      ? std::distance(squeeze_dim_flags.begin(), it)
+      : -1;
+}
+
+// If the def of lookUpTv is squeeze, and we're moving the squeeze to
+// after the gather/take_along_axis, then we need to add an unsqueeze
+// to the IndexTv.
+// The def in the arugment is the Expr that define the lookupTv for the
+// original take_along_axis.
+TensorView* prepareIndexTv(GatherOp* gather_op, Expr* def) {
+  if (def->isA<SqueezeOp>()) {
+    auto squeeze_dim_flags = def->as<SqueezeOp>()->getSqueezeDimFlags();
+    auto dim = positionOfSqueezedDim(squeeze_dim_flags);
+    return unsqueeze(gather_op->indexTv(), dim);
+  }
+  return gather_op->indexTv();
+}
+
+// def is the Expr the defines the lookupTv for take_along_axis.
+TensorView* addPostGatherSqueeze(
+    GatherOp* old_gather,
+    GatherOp* new_gather,
+    Expr* def) {
+  auto new_squeeze_output = squeeze(
+      new_gather->output(0)->as<TensorView>(),
+      def->as<SqueezeOp>()->getSqueezeDimFlags());
+
+  TransformReplay::selfReplay(
+      old_gather->output(0)->as<TensorView>()->domain(),
+      new_squeeze_output->domain());
+
+  return new_squeeze_output;
+}
+
+// def is the Expr the defines the lookupTv for take_along_axis.
+TensorView* addPostGatherOp(
```

Comment:
nitpick: this is really `addPostGatherUnary`. just to be consistent with `addPostGatherSqueeze` above.